crates/ruma-events/src/room/name.rs

```diff
@@ -11,51 +13,78 @@ pub type NameEvent = StateEvent<NameEventContent>;
 /// The payload for `NameEvent`.
 #[derive(Clone, Debug, Deserialize, Serialize, EventContent)]
 #[ruma_event(type = "m.room.name", kind = State)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
 pub struct NameEventContent {
-    /// The name of the room. This MUST NOT exceed 255 bytes.
-    #[serde(default, deserialize_with = "room_name")]
-    pub(crate) name: Option<String>,
+    /// The name of the room.
+    #[serde(default, deserialize_with = "ruma_serde::empty_string_as_none")]
+    pub name: Option<RoomName>,
 }
 
 impl NameEventContent {
     /// Create a new `NameEventContent` with the given name.
-    ///
-    /// # Errors
-    ///
-    /// `InvalidInput` will be returned if the name is more than 255 bytes.
-    pub fn new(name: String) -> Result<Self, InvalidInput> {
-        match name.len() {
-            0 => Ok(Self { name: None }),
-            1..=255 => Ok(Self { name: Some(name) }),
-            _ => Err(InvalidInput("a room name cannot be more than 255 bytes".into())),
-        }
+    pub fn new(name: Option<RoomName>) -> Self {
+        Self { name }
     }
 
     /// The name of the room, if any.
-    pub fn name(&self) -> Option<&str> {
-        self.name.as_deref()
+    pub fn name(&self) -> Option<&RoomName> {
+        self.name.as_ref()
     }
 }
 
-fn room_name<'de, D>(deserializer: D) -> Result<Option<String>, D::Error>
-where
-    D: serde::de::Deserializer<'de>,
-{
-    use serde::de::Error;
-
-    // this handles the null case and the empty string or nothing case
-    match Option::<String>::deserialize(deserializer)? {
-        Some(name) => match name.len() {
-            0 => Ok(None),
-            1..=255 => Ok(Some(name)),
-            _ => Err(D::Error::custom("a room name cannot be more than 255 bytes")),
-        },
-        None => Ok(None),
+/// The name of a room.
+///
+/// It should not exceed 255 characters and should not be empty.
+#[derive(Clone, Debug, Serialize, PartialEq, Eq)]
+#[serde(transparent)]
+pub struct RoomName(String);
+
+impl TryFrom<String> for RoomName {
+    type Error = InvalidInput;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        match value.len() {
+            0 => Err(InvalidInput("a room name cannot be empty.".into())),
+            1..=255 => Ok(RoomName(value)),
+            _ => Err(InvalidInput("a room name cannot be more than 255 bytes.".into())),
+        }
+    }
+}
+
+impl From<RoomName> for String {
+    fn from(name: RoomName) -> Self {
+        name.0
+    }
+}
+
+impl<'de> Deserialize<'de> for RoomName {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        let str_name = String::deserialize(deserializer)?;
+
+        match RoomName::try_from(str_name) {
+            Ok(name) => Ok(name),
+            Err(e) => Err(D::Error::custom(e.to_string())),
+        }
+    }
+}
+
+impl Deref for RoomName {
```

Comment:
Thinking more about it, since we don't have `Deref` impls for the types in ruma-identifiers, let's not have on here either for now.